### PR TITLE
Add C# subscription Maestro scenario tests

### PR DESCRIPTION
**Pull request by meganaquinn**
Also adds option to use stdin with Darc commands and implements some of the commands for use with the cloning verification test.

### CODE REVIEW

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 1]
```diff
@@ -85,13 +86,40 @@ public string GetRepoFetchUrl(string org, string repository)
...
+                "--github-pat", _parameters.GitHubToken,
+                "--azdev-pat", _parameters.AzDoToken,
+            }).ToArray());
+        }
+
         public Task<string> RunDarcAsync(params string[] args)
         {
             return TestHelpers.RunExecutableAsync(_parameters.DarcExePath, args.Concat(new[]
-            {
+{
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:118]: This is a little weird lookin'.

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 2]
```diff
@@ -153,14 +181,50 @@ public async Task DeleteDefaultTestChannelAsync(string testChannelName, string r
             await RunDarcAsync("delete-default-channel", "--channel", testChannelName, "--repo", repoUri, "--branch", branch).ConfigureAwait(false);
         }
 
-        public async Task<AsyncDisposableValue<string>> CreateSubscriptionAsync(string sourceChannelName, string sourceRepo, string targetRepo, string targetBranch, string updateFrequency)
+        public async Task<AsyncDisposableValue<string>> CreateSubscriptionAsync(string sourceChannelName, string sourceRepo, string targetRepo, string targetBranch, string updateFrequency, string sourceOrg = "dotnet",
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:184]: Can we chop this (put one parameter per line).  It's hard to read like this (and the first line is too long).

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 3]
```diff
@@ -153,14 +181,50 @@ public async Task DeleteDefaultTestChannelAsync(string testChannelName, string r
...
+                        await RunDarcAsync("delete-subscriptions", "--id", subscriptionId, "--quiet").ConfigureAwait(false);
+                    }
+                    catch (MaestroTestException)
+                    {
+                        // If this throws an exception the most likely cause is that the subscription was deleted as part of the test case
+                    }
+                });
+            }
+
+            throw new MaestroTestException("Unable to create subscription.");
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:222]: Wait... this method _always_ throws.  There are no returns in this method, and this "throw" isn't in any sort of statement.

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 4]
```diff
@@ -169,13 +233,61 @@ public async Task<AsyncDisposableValue<string>> CreateSubscriptionAsync(string s
...
+            }
+            else
+            {
+                await RunDarcAsync("subscription-status", "--id", subscriptionId, actionToTake, "--quiet").ConfigureAwait(false);
+            }
+        }
+
+        public async Task<string> DeleteSubscriptionsForChannel(string channelName)
+        {
+            string[] parameters = { "delete-subscriptions", "--channel", $"\"{channelName}\"", "--quiet" };
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:282]: Nothing uses this array.

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 5]
```diff
@@ -169,13 +233,61 @@ public async Task<AsyncDisposableValue<string>> CreateSubscriptionAsync(string s
...
+            return await RunDarcAsync("delete-subscriptions", "--channel", $"{channelName}", "--quiet");
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:283]: No need to interpolate a single string variable, just pass it

```suggestion
            return await RunDarcAsync("delete-subscriptions", "--channel", channelName, "--quiet");
```

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 6]
```diff
@@ -280,6 +392,20 @@ public async Task<TemporaryDirectory> CloneRepositoryAsync(string org, string re
             return shareable.TryTake()!;
         }
 
+        public async Task DarcCloneRepositoryAsync(string sourceRepoUri, string sourceRepoVersion, string gitDirFolder,
+            string reposToIgnore, string reposFolder, int depth, bool includeToolset)
+        {
+            string[] parameters = {"clone", "--repo", sourceRepoUri, "--version", sourceRepoVersion, "--git-dir-folder",
+               gitDirFolder, "--ignore-repos", reposToIgnore, "--repos-folder", reposFolder, "--depth", depth.ToString() };
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:399]:
```suggestion
            string[] parameters = {
                "clone",
                "--repo", sourceRepoUri,
                "--version", sourceRepoVersion,
                "--git-dir-folder", gitDirFolder,
                "--ignore-repos", reposToIgnore,
                "--repos-folder", reposFolder,
                "--depth", depth.ToString()
            };
```

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 7]
```diff
@@ -323,5 +449,87 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
...
+                string path = Path.Combine(reposFolder, name);
+                bool checkDir = Directory.Exists(path);
+                Assert.IsTrue(Directory.Exists(path), $"Expected cloned repo '{name}' but not found at '{path}'");
+
+                string versionDetailsPath = Path.Combine(path, "eng", "Version.Details.xml");
+                if (File.Exists(versionDetailsPath))
+                {
+                    using (FileStream stream = File.OpenRead(versionDetailsPath))
+                    {
+                        SHA256 Sha256 = SHA256.Create();
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:483]: I think this is IDisposable too.

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 8]
```diff
@@ -323,5 +449,87 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
...
+                        byte[] hash = Sha256.ComputeHash(stream);
+                        string hashString = BitConverter.ToString(hash);
+                        Assert.AreEqual(hashString, repos[name], $"Expected {versionDetailsPath} to have hash '{repos[name]}', actual hash '{hashString}'");
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:486]: We are... hashing the Versions file? That's going to make the failure really hard to debug. Though I don't see any calls to this method...

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 9]
```diff
@@ -323,5 +449,87 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
...
+
+            foreach (string repo in masterRepos)
+            {
+                string repoPath = Path.Combine(reposFolder, repo);
+                Assert.IsTrue(Directory.Exists(repoPath), $"Expected cloned master repo '{repo}' but not found at '{repoPath}'");
+
+                string gitRedirectPath = Path.Combine(repoPath, ".git");
+                string expectedGitDir = Path.Combine(gitDirFolder, repoPath);
+                string expectedRedirect = $"gitdir: {expectedGitDir}.git";
+                string actualRedirect = Directory.GetFiles(gitRedirectPath).First();
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:503]: Why call "GetFiles" when we have the full path?  File.Exists(gitRedirectPath) seems like a clearer way to do this...  Otherwise this is going to throw a `ArgumentException("source is empty")`

File: src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionStringBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
+using System.Text;
+using Microsoft.DotNet.Darc;
+using Microsoft.DotNet.Maestro.Client.Models;
+
+namespace Maestro.ScenarioTests.ObjectHelpers
+{
+    /// <summary>
+    /// Contains the data for a subscription object and the logic for formatting into a darc return string, based on the code in Darc.GetSubscriptionsOperation
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionStringBuilder.cs:8]: Can we just use that code rather than re-implement it here?  Otherwise if we ever touch it, we need to come here and change it here too.

File: src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+                    await using AsyncDisposableValue<string> subscription2Id = await CreateSubscriptionAsync(
+                        channel1Name, repo1Name, repo1Name, targetBranch, "none", "maestro-auth-test",
+                        new List<string>
+                        { "--all-checks-passed", "--no-extra-commits", "--no-requested-changes", "--ignore-checks", "WIP,license/cla" }, targetIsAzDo: true);
+
+                    Subscription expectedSubscription2 = new Subscription(Guid.Parse(subscription2Id.Value), true, repo1Uri, repo1AzDoUri, targetBranch);
+                    expectedSubscription2.Channel = new Channel(42, channel1Name, "test");
+                    expectedSubscription2.Policy = new SubscriptionPolicy(false, UpdateFrequency.None);
+
+                    string expectedPoliciesSub2 = "  - Merge Policies:\r\n    NoExtraCommits\r\n    AllChecksSuccessful\r\n      ignoreChecks = \r\n" +
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs:75]: Use a verbatim string, instead of a bunch of normal string with new lines embedded an contatenated together.

File: src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+        [TearDown]
+        public Task DisposeAsync()
+        {
+            _parameters.Dispose();
+            return Task.CompletedTask;
+        }
+
+        [Test]
+        public async Task Subscriptions_EndToEnd()
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs:37]: Wow... this test is REALLY long, and doesn't assert much... It does a lot of actions, doesn't check that they did what it expects... only way at the end that somewhere, in this 200 lines of tests, something didn't do what it was supposed to. :-(

This isn't your fault... the PS tests do this too, but it makes it basically impossible to tell why the tests are failing. :-(

File: src/Maestro/tests/Maestro.ScenarioTests/TestHelpers.cs [hunk 13]
```diff
@@ -40,15 +48,17 @@ void WriteOutput(string message)
...
                 StartInfo = psi
             };
+
             var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
             process.EnableRaisingEvents = true;
             process.Exited += (s, e) => { tcs.TrySetResult(true); };
             process.Start();
 
             Task<bool> exitTask = tcs.Task;
+            Task stdin = Task.Run(async () => { process.StandardInput.Write(input); process.StandardInput.Close(); });
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/TestHelpers.cs:58]: I'm... confused about this call. There isn't anything async in here... Are we not able to just call "process.StandardInput.Write(input)" here without the need for the Task.Run?

File: src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs [hunk 14]
```diff
@@ -155,11 +155,10 @@ protected void Clone(string repoUri, string commit, string targetDirectory, ILog
                     Password = pat
                 },
             };
-            using (_logger.BeginScope("Cloning {repoUri} to {targetDirectory}", repoUri, targetDirectory))
+            using (_logger.BeginScope($"Cloning {repoUri} to {targetDirectory}"))
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs:158]: Why this change?

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 15]
```diff
@@ -85,13 +86,40 @@ public string GetRepoFetchUrl(string org, string repository)
...
+            {
+                azdoProject = "internal";
+            }
+
+            return $"https://dev.azure.com/{azdoAccount}/{azdoProject}/_git/{repoName}";
+        }
+
+        public Task<string> RunDarcAsyncWithInput(string input, params string[] args)
+        {
+            return TestHelpers.RunExecutableAsyncWithInput(_parameters.DarcExePath, input, args.Concat(new[]
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:106]: It's... interesting that we are importing the Darc project... we could just... call the operations we want rather than having to build up the giant exe command line.

Comment (ChadNedzlek) [reply src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:106]: That would also mean we didn't have to do all the string parsing on the output of the process. :-)

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 16]
```diff
@@ -169,13 +233,61 @@ public async Task<AsyncDisposableValue<string>> CreateSubscriptionAsync(string s
...
         }
 
+        public async Task<string> GetSubscriptionInfo(string subscriptionId)
+        {
+            return await RunDarcAsync("get-subscriptions", "--ids", subscriptionId).ConfigureAwait(false);
+        }
+
+        public async Task<string> GetSubscriptions(string channelName)
+        {
+            return await RunDarcAsync("get-subscriptions", "--channel", $"\"{channelName}\"");
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:257]: We shouldn't have to have these quotes here... this implies RunDarcAsync isn't quoting parameters correctly itself.

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 17]
```diff
@@ -323,5 +449,87 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
...
+                    {
+                        SHA256 Sha256 = SHA256.Create();
+                        byte[] hash = Sha256.ComputeHash(stream);
+                        string hashString = BitConverter.ToString(hash);
+                        Assert.AreEqual(hashString, repos[name], $"Expected {versionDetailsPath} to have hash '{repos[name]}', actual hash '{hashString}'");
+                    }
+                }
+                else
+                {
+                    Assert.AreEqual(repos[name], "", $"Expected a '{versionDetailsPath}' but it is missing");
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:491]: It seems like this is supposed to be FileAssert.Exists(versionDetailsFile) back up on line 479, then ditch the if.

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs (hunk 3 above)
Comment (meganaquinn) [reply src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:222]: The return is on line 208

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 18]
```diff
@@ -323,5 +449,87 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
...
+            foreach (string name in repos.Keys)
+            {
+                string path = Path.Combine(reposFolder, name);
+                bool checkDir = Directory.Exists(path);
+                Assert.IsTrue(Directory.Exists(path), $"Expected cloned repo '{name}' but not found at '{path}'");
+
+                string versionDetailsPath = Path.Combine(path, "eng", "Version.Details.xml");
+                if (File.Exists(versionDetailsPath))
+                {
+                    using (FileStream stream = File.OpenRead(versionDetailsPath))
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:481]: NUnit has a "FileAssert.AreEqual", so we could just check the file to be the right value.

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 19]
```diff
@@ -323,5 +449,87 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
...
+                else
+                {
+                    Assert.AreEqual(repos[name], "", $"Expected a '{versionDetailsPath}' but it is missing");
+                }
+            }
+
+            foreach (string repo in masterRepos)
+            {
+                string repoPath = Path.Combine(reposFolder, repo);
+                Assert.IsTrue(Directory.Exists(repoPath), $"Expected cloned master repo '{repo}' but not found at '{repoPath}'");
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:498]: There is a DirectoryAsserts.Exists(repoPath)

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 20]
```diff
@@ -323,5 +449,87 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
...
+
+                string gitRedirectPath = Path.Combine(repoPath, ".git");
+                string expectedGitDir = Path.Combine(gitDirFolder, repoPath);
+                string expectedRedirect = $"gitdir: {expectedGitDir}.git";
+                string actualRedirect = Directory.GetFiles(gitRedirectPath).First();
+                StringAssert.AreEqualIgnoringCase(expectedRedirect, actualRedirect, $"Expected '{repoPath}' to have a .gitdir redirect of '{expectedRedirect}', actual '{actualRedirect}'");
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:504]: Is this trying to compare the contents of the file?  Right now it's just checking the file name to a thing with a colon in it, so it's just always going to file (file names can't have colons in them). Does darc mess with .gitdir stuff... I'm not sure what this assert is actually testing, and what it means if it fails. :-(

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs [hunk 21]
```diff
@@ -323,5 +449,87 @@ public async Task<string> GetRepositoryPolicies(string repoUri, string branchNam
...
+            {
+                string gdPath = Path.Combine(gitDirFolder, gitDir);
+                Assert.IsTrue(Directory.Exists(gdPath), $"Expected a .gitdir for '{gitDir}' but not found at '{gdPath}'");
+            }
+
+            string[] actualGitDirs = Directory.GetDirectories(gitDirFolder);
+
+            foreach (string gitDir in actualGitDirs)
+            {
+                Assert.Contains(gitDir, actualGitDirs, $"Found unexpected .gitdir '{gitDir}'");
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:530]: ... This is asserting that the thing we are iterating over contains the things inside it...

File: src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+using Maestro.ScenarioTests.ObjectHelpers;
+using Microsoft.DotNet.Maestro.Client.Models;
+using NUnit.Framework;
+using NUnit.Framework.Internal;
+
+namespace Maestro.ScenarioTests
+{
+    public class ScenarioTests_Subscriptions : MaestroScenarioTestBase
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs:12]: Can you put the [Category="ScenarioTests"] that I put on the other scenario tests? It will help make sure we only run stuff at the right time if we ever have other assemblies.

File: src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+                    expectedYamlSubscription2.Channel = new Channel(42, channel1Name, "test");
+                    expectedYamlSubscription2.Policy = new SubscriptionPolicy(false, UpdateFrequency.EveryWeek);
+                    expectedYamlSubscription2.Policy.MergePolicies = ImmutableList.Create(standardPolicy);
+
+                    string expectedYamlSubscriptionInfo2 = SubscriptionStringBuilder.GetSubscriptionString(expectedYamlSubscription2, "  - Merge Policies:\r\n    standard");
+
+                    await ValidateSubscriptionInfo(yamlSubscription2Id.Value, expectedYamlSubscriptionInfo2);
+                    await DeleteSubscriptionById(yamlSubscription2Id.Value);
+
+                    TestContext.WriteLine("Attempt to add multiple of the same merge policy checks. Should fail.");
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs:196]: This "test" appears to actually be a lot of tests mushed into one.   There is an "IgnoresCase" test, a "MultipleMergePolicyRejected" test... can we break these up so that the tests are a touch more granular than "something went wrong somewhere"?

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs (hunk 3 above)
Comment (ChadNedzlek) [reply src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:222]: Oh wow... I didn't notice that.

Can we swap the order... so if it says `if (!match.Succes) { throw }`?

Technically just because that string isn't there, doesn't mean we don't need to clean up the subscription.  I'm not sure how to handle that without the tests leaking subscriptions all over the place.

File: src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs (hunk 14 above)
Comment (alexperovich) [reply src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs:158]: This is actually a functional change. We no longer have the key-value data in the scope. This should be reverted.

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs (hunk 8 above)
Comment (meganaquinn) [reply src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:486]: This is what's done in the repo cloning powershell test, which is still in progress. I'm absolutely open to better ideas for any of the cloning stuff, it's fairly murky in a lot of places.

File: src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs (hunk 12 above)
Comment (meganaquinn) [reply src/Maestro/tests/Maestro.ScenarioTests/ScenarioTests_Subscriptions.cs:37]: I absolutely agree. I'd like to split it into individual tests, but I think that's going to get messy with even more things running in parallel (especially with the subscription tests, which use the same 3 repos for everything & will fail on duplicates) and I'd rather have some version of all the powershell tests checked in than spend a bunch more time refactoring.

File: src/Maestro/tests/Maestro.ScenarioTests/TestHelpers.cs (hunk 13 above)
Comment (meganaquinn) [reply src/Maestro/tests/Maestro.ScenarioTests/TestHelpers.cs]: I pulled the async (it was extraneous in any case), but this is the set up that Alex recommended. I'll try it without a Task, but my vague understanding was that the task is needed because of the way the output is handled.

File: src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs (hunk 14 above)
Comment (meganaquinn) [reply src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs:158]: I honestly can't remember, I think it wasn't printing properly when I was debugging (I can't think of any other reason I would have been in that code at all). I'll revert it and remind myself what was going on.

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs (hunk 15 above)
Comment (meganaquinn) [reply src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:106]: I really like that idea, but would it reduce the code coverage we want? Right now I'm explicitly only using it to build expected results, not to do any of the test calls.

Comment (ChadNedzlek) [reply src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs:106]: Makes sense. Let get it in as is for now, and we can think about changes later (that probably goes for a lot of my comments)

File: src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs (hunk 14 above)
Comment (meganaquinn) [reply src/Microsoft.DotNet.Darc/src/DarcLib/RemoteRepoBase.cs:158]: Done.

File: src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionStringBuilder.cs (hunk 10 above)
Comment (meganaquinn) [reply src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionStringBuilder.cs]: I have a new version with this change - it would required refactoring the production function some though, so I definitely want to get a bunch of eyes on it to make sure I didn't inadvertently change something (based on my local testing there is no functional change, just a new function).

File: src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs (hunk 20 above)
Comment (meganaquinn) [reply src/Maestro/tests/Maestro.ScenarioTests/MaestroScenarioTestBase.cs]: I'm missing information about how the cloning test is supposed to work and what it's supposed to be covering, so I pulled the code from this PR. I'm going to follow up with Ricardo this week and see if he has the missing pieces for me, then I'll put out another PR with this test.

File: src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionBuilder.cs [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+        public static Subscription BuildSubscription(string repo1Uri, string repo2Uri, string targetBranch, string channelName, string subscriptionId,
+            UpdateFrequency updateFrequency, bool batchable, List<string> mergePolicyNames, List<string> ignoreChecks)
+        {
+            Subscription expectedSubscription = new Subscription(Guid.Parse(subscriptionId), true, repo1Uri, repo2Uri, targetBranch);
+            expectedSubscription.Channel = new Channel(42, channelName, "test");
+            expectedSubscription.Policy = new SubscriptionPolicy(batchable, updateFrequency);
+
+            List<MergePolicy> mergePolicies = new List<MergePolicy>();
+
+            if (mergePolicyNames != null)
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionBuilder.cs:25]: I'd invert this if.  Right now the else is so far away that it's hard to keep it all in my head to figure out what it goes with, and what the behavior means.  In general, every level of indentation is some piece of context a reader/dev has to keep in their head when looking at the code.  So "less indented" translates to "less cognitive load" someone has to have to understand the code.  Less load means less likely to miss something and introduce a bug.

Also... I don't see where the "else" path returns, so I'm a little confused how this compiles...

File: src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionBuilder.cs [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+                    mergePolicies.Add(
+                        new MergePolicy
+                        {
+                            Name = Constants.AllCheckSuccessfulMergePolicyName,
+                            Properties = ImmutableDictionary.Create<string, JToken>()
+                                .Add(Constants.IgnoreChecksMergePolicyPropertyName, JToken.FromObject(ignoreChecks))
+                        });
+                }
+
+                if (mergePolicyNames.Contains(Constants.NoRequestedChangesMergePolicyName))
```
Comment (ChadNedzlek) [inline src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionBuilder.cs:56]: Are the "merge policys" a fixed, enumerable set?  If so... it should probably be an enumeration rather than a bunch of strings.  And if we can have more than once, then it can be a "[Flags]" enum rather than a List<MergePolicies>.  It seems like the are based on this code, and then you don't have to test for what happens if some passed "Pizza" into this method. :-D

Comment (meganaquinn) [reply src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionBuilder.cs:56]: To clarify, are you suggesting adding the enum for the test code or in Darc? Darc uses the strings & constants.

Comment (ChadNedzlek) [reply src/Maestro/tests/Maestro.ScenarioTests/ObjectHelpers/SubscriptionBuilder.cs]: Well, really I'd like darc not to use a bunch of magic strings.  But that's probably a bigger change.  We should probably log a bug for it that can get tossed into the void. :-)

**Comment by meganaquinn** [comment]
Abandoning this in favor of a branch that's clean (re-basing made this unreadable). 
https://github.com/dotnet/arcade-services/pull/1216

**Comment by ChadNedzlek** [comment]
That's why I never rebase.  Usually just merging in master is easier, and produces easier to deal with commit history. :-)
